akka-stream/src/main/scala/akka/stream/Attributes.scala

```diff
@@ -201,6 +150,99 @@ final case class Attributes(attributeList: List[Attributes.Attribute] = Nil) {
+  }
+
+  /**
+   * Java API: Get all attributes of a given `Class` or
+   * subclass thereof.
+   *
+   * The list is ordered with the most specific attribute first, least specific last.
+   * Note that the order was the opposite in Akka 2.4.x.
+   *
+   * Note that stages in general should not inspect the whole hierarchy but instead use
+   * `get` to get the most specific attribute value.
+   */
+  def getAttributeList[T <: Attribute](c: Class[T]): java.util.List[T] =
+    if (attributeList.isEmpty) java.util.Collections.emptyList()
+    else {
+      val result = new java.util.ArrayList[T]
+      attributeList.foreach { a ⇒
+        if (c.isInstance(a))
+          result.add(c.cast(a))
+      }
+      result
+    }
+
+  /**
+   * Scala API: Get all attributes of a given type (or subtypes thereof).
+   *
+   * Note that stages in general should not inspect the whole hierarchy but instead use
+   * `get` to get the most specific attribute value.
+   *
+   * The list is ordered with the most specific attribute first, least specific last.
+   * Note that the order was the opposite in Akka 2.4.x.
+   */
+  def filtered[T <: Attribute: ClassTag]: List[T] = {
+    val c = implicitly[ClassTag[T]].runtimeClass.asInstanceOf[Class[T]]
+    attributeList.collect { case attr if c.isAssignableFrom(attr.getClass) ⇒ c.cast(attr) }
+  }
+
+  /**
+   * Java API: Get the least specific attribute (added first) of a given `Class` or subclass thereof.
+   * If no such attribute exists the `default` value is returned.
+   */
+  @deprecated("Attributes should always be most specific, use getAttribute[T]", "2.5.7")
+  def getFirstAttribute[T <: Attribute](c: Class[T], default: T): T =
+    getFirstAttribute(c).orElse(default)
+
+  /**
+   * Java API: Get the least specific attribute (added first) of a given `Class` or subclass thereof.
+   */
+  @deprecated("Attributes should always be most specific, use get[T]", "2.5.7")
+  def getFirstAttribute[T <: Attribute](c: Class[T]): Optional[T] =
+    attributeList.reverseIterator.collectFirst { case attr if c.isInstance(attr) ⇒ c cast attr }.asJava
+
+  /**
+   * Scala API: Get the least specific attribute (added first) of a given type parameter T `Class` or subclass thereof.
+   * If no such attribute exists the `default` value is returned.
+   */
+  @deprecated("Attributes should always be most specific, use get[T]", "2.5.7")
+  def getFirst[T <: Attribute: ClassTag](default: T): T = {
+    getFirst[T] match {
+      case Some(a) ⇒ a
+      case None    ⇒ default
+    }
+  }
+
+  /**
+   * Scala API: Get the least specific attribute (added first) of a given type parameter T `Class` or subclass thereof.
+   */
+  @deprecated("Attributes should always be most specific, use get[T]", "2.5.7")
+  def getFirst[T <: Attribute: ClassTag]: Option[T] = {
+    val c = classTag[T].runtimeClass.asInstanceOf[Class[T]]
+    attributeList.reverseIterator.collectFirst { case attr if c.isInstance(attr) ⇒ c.cast(attr) }
+  }
+
```

Comment:
> As addAttributes is implemented in terms of withAttributes (which I think is needed for the types to be right) we can't remove withAttributes
> Maybe we could deprecate and make private/protected in the future though.

I think it would be good to deprecate it and later make it internal.

Reply:
good idea

Reply:
Hah, ofc it was a good idea, it was my idea! 👯‍♂️

Reply:
Argh, just found a problem with this, since add adds rather than replaces, this will break the whole idea of concatenating names into a "a.b.c" structure - every stage already has a name, and add will just append onto that while `withAttributes()` will actually replace it (together with all other attributes on the same `Graph`).

Reply:
@johanandren what is the status here? Can't we do the right thing in `.named`?

Reply:
Maybe, I'll try that and see.